{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 454,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import collections\n",
    "import skbio \n",
    "from skbio.diversity.alpha import shannon \n",
    "import IPython\n",
    "import re\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Eerst alle stappen uitvoeren met de data van de VDJdb (gedownload op 11 december), splitsen + weergeven van data\n",
    "vdjdb_data = pd.read_csv(\"vdjdb_data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>Gene</th>\n",
       "      <th>J</th>\n",
       "      <th>V</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAVGNNAGNMLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAJ39*01</td>\n",
       "      <td>TRAV20*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAAKTGGGNKLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAJ10*01</td>\n",
       "      <td>TRAV29/DV5*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAAKAGGGNKLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAJ10*01</td>\n",
       "      <td>TRAV29/DV5*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAAITGIGGSQGNLIF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAJ42*01</td>\n",
       "      <td>TRAV13-1*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAGTDGGATNKLIF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAJ32*01</td>\n",
       "      <td>TRAV36/DV7*01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               CDR3   Epitope Gene          J              V\n",
       "0     CAVGNNAGNMLTF  NEGVKAAW  TRA  TRAJ39*01      TRAV20*01\n",
       "1     CAAKTGGGNKLTF  NEGVKAAW  TRA  TRAJ10*01  TRAV29/DV5*01\n",
       "2     CAAKAGGGNKLTF  NEGVKAAW  TRA  TRAJ10*01  TRAV29/DV5*01\n",
       "3  CAAITGIGGSQGNLIF  NEGVKAAW  TRA  TRAJ42*01    TRAV13-1*01\n",
       "4    CAGTDGGATNKLIF  NEGVKAAW  TRA  TRAJ32*01  TRAV36/DV7*01"
      ]
     },
     "execution_count": 456,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del vdjdb_data ['Unnamed: 0']\n",
    "vdjdb_data = vdjdb_data[sorted(vdjdb_data.columns)]\n",
    "vdjdb_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {},
   "outputs": [],
   "source": [
    "TRAJ = vdjdb_data ['J'].unique()\n",
    "TRAV = vdjdb_data ['V'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TRAJ39*01' 'TRAJ10*01' 'TRAJ42*01' 'TRAJ32*01' 'TRAJ13*01' 'TRAJ8*01'\n",
      " 'TRAJ47*01' 'TRAJ29*01' 'TRAJ17*01' 'TRAJ38*01' 'TRAJ44*01' 'TRAJ49*01'\n",
      " 'TRAJ26*01' 'TRAJ34*01' 'TRAJ40*01' 'TRAJ33*01' 'TRAJ23*01' 'TRAJ57*01'\n",
      " 'TRAJ43*01' 'TRAJ6*01' 'TRAJ20*01' 'TRAJ30*01' 'TRAJ48*01' 'TRAJ31*01'\n",
      " 'TRAJ45*01' 'TRAJ3*01' 'TRAJ27*01' 'TRAJ36*01' 'TRAJ52*01' 'TRAJ15*01'\n",
      " 'TRAJ5*01' 'TRAJ4*01' 'TRAJ50*01' 'TRAJ28*01' 'TRAJ53*01' 'TRAJ37*01'\n",
      " 'TRAJ22*01' 'TRAJ41*01' 'TRAJ9*01' 'TRAJ12*01' 'TRAJ11*01' 'TRAJ46*01'\n",
      " 'TRAJ54*01' 'TRAJ18*01' 'TRAJ21*01' 'TRAJ56*01' 'TRAJ7*01' 'TRAJ14*01'\n",
      " 'TRAJ13*02' 'TRAJ24*02']\n"
     ]
    }
   ],
   "source": [
    "print (TRAJ) # via deze methode gekeken hoe de data eruitzag om te splitsen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TRAV20*01' 'TRAV29/DV5*01' 'TRAV13-1*01' 'TRAV36/DV7*01' 'TRAV19*01'\n",
      " 'TRAV8-2*01' 'TRAV27*01' 'TRAV8-6*01' 'TRAV35*01' 'TRAV9-2*01'\n",
      " 'TRAV23/DV6*01' 'TRAV38-2/DV8*01' 'TRAV18*01' 'TRAV8-3*01' 'TRAV22*01'\n",
      " 'TRAV13-2*01' 'TRAV21*01' 'TRAV8-4*01' 'TRAV5*01' 'TRAV12-2*01'\n",
      " 'TRAV38-1*01' 'TRAV12-1*01' 'TRAV1-2*01' 'TRAV39*01' 'TRAV16*01'\n",
      " 'TRAV17*01' 'TRAV10*01' 'TRAV24*01' 'TRAV41*01' 'TRAV12-3*01' 'TRAV3*01'\n",
      " 'TRAV8-1*01' 'TRAV26-2*01' 'TRAV4*01' 'TRAV14/DV4*01' 'TRAV2*01'\n",
      " 'TRAV30*01' 'TRAV25*01' 'TRAV26-1*01' 'TRAV1-1*01' 'TRAV34*01'\n",
      " 'TRAV40*01' 'TRAV6*01' 'TRAV9-1*01' 'TRAV7*01' 'TRAV14/DV4*02'\n",
      " 'TRAV8-6*02']\n"
     ]
    }
   ],
   "source": [
    "print (TRAV) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>Gene</th>\n",
       "      <th>V</th>\n",
       "      <th>J_family</th>\n",
       "      <th>J_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAVGNNAGNMLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>39</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAAKTGGGNKLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV29/DV5*01</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAAKAGGGNKLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV29/DV5*01</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAAITGIGGSQGNLIF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV13-1*01</td>\n",
       "      <td>42</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAGTDGGATNKLIF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV36/DV7*01</td>\n",
       "      <td>32</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               CDR3   Epitope Gene              V J_family J_allele\n",
       "0     CAVGNNAGNMLTF  NEGVKAAW  TRA      TRAV20*01       39       01\n",
       "1     CAAKTGGGNKLTF  NEGVKAAW  TRA  TRAV29/DV5*01       10       01\n",
       "2     CAAKAGGGNKLTF  NEGVKAAW  TRA  TRAV29/DV5*01       10       01\n",
       "3  CAAITGIGGSQGNLIF  NEGVKAAW  TRA    TRAV13-1*01       42       01\n",
       "4    CAGTDGGATNKLIF  NEGVKAAW  TRA  TRAV36/DV7*01       32       01"
      ]
     },
     "execution_count": 460,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# splitsen van de J-kolom, eerst wordt TRAJ verwijderd, daarna zal het allel worden afgesplitst en in een andere kolom worden geplaatst, de oorspronkelijke kolom zal worden verwijderd en de data wordt aangevuld zodat zowel de family als het allele steeds 2 digits heeft\n",
    "vdjdb_data.replace('TRAJ','',regex=True, inplace = True)\n",
    "vdjdb_data[['J_family','J_allele']] = vdjdb_data.J.str.split (\"*\",expand=True)\n",
    "del vdjdb_data ['J']\n",
    "# add a leading 0 to all families with only one digit\n",
    "vdjdb_data['J_family'] = vdjdb_data['J_family'].str.zfill(2)\n",
    "vdjdb_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Splitsen van de V kolom, eerst zal de TRAV worden verwijderd en zal terug het allele worden afgesplitst en in een nieuwe kolom worden geplaatst\n",
    "# daarna de gene van de family scheiden \n",
    "vdjdb_data.replace('TRAV','',regex=True, inplace = True)\n",
    "vdjdb_data[['V','V_allele']] = vdjdb_data.V.str.split (\"*\",expand=True) # eerst allele afsplitsen, komt na de *\n",
    "# dan in 3 stappen de gene van de family scheiden\n",
    "vdjdb_data[['V_family','V_gene_1']] = vdjdb_data.V.str.split (\"-\",expand=True) # eerst scheiden op -\n",
    "vdjdb_data[['V_family','V_gene_2']] = vdjdb_data.V_family.str.split (\"/\",expand=True) # dan op /\n",
    "vdjdb_data.replace(np.nan,'', regex=True )\n",
    "# ten slotte de 2 gene kolommen nog mergen en de andere 2 verwijderen\n",
    "vdjdb_data = vdjdb_data.assign(V_gene = vdjdb_data.V_gene_1.astype(str) + vdjdb_data.V_gene_2.astype(str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {},
   "outputs": [],
   "source": [
    "vdjdb_data.replace('nan','',regex=True, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "metadata": {},
   "outputs": [],
   "source": [
    "del vdjdb_data ['V']\n",
    "del vdjdb_data ['V_gene_1']\n",
    "del vdjdb_data ['V_gene_2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a leading 0 to all Values with only one digit\n",
    "vdjdb_data['V_family'] = vdjdb_data['V_family'].str.zfill(2)\n",
    "vdjdb_data['V_gene'] = vdjdb_data['V_gene'].str.zfill(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>Gene</th>\n",
       "      <th>J_allele</th>\n",
       "      <th>J_family</th>\n",
       "      <th>V_allele</th>\n",
       "      <th>V_family</th>\n",
       "      <th>V_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAVGNNAGNMLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>01</td>\n",
       "      <td>39</td>\n",
       "      <td>01</td>\n",
       "      <td>20</td>\n",
       "      <td>00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAAKTGGGNKLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>29</td>\n",
       "      <td>DV5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAAKAGGGNKLTF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>01</td>\n",
       "      <td>10</td>\n",
       "      <td>01</td>\n",
       "      <td>29</td>\n",
       "      <td>DV5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAAITGIGGSQGNLIF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>01</td>\n",
       "      <td>42</td>\n",
       "      <td>01</td>\n",
       "      <td>13</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAGTDGGATNKLIF</td>\n",
       "      <td>NEGVKAAW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>01</td>\n",
       "      <td>32</td>\n",
       "      <td>01</td>\n",
       "      <td>36</td>\n",
       "      <td>DV7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               CDR3   Epitope Gene J_allele J_family V_allele V_family V_gene\n",
       "0     CAVGNNAGNMLTF  NEGVKAAW  TRA       01       39       01       20     00\n",
       "1     CAAKTGGGNKLTF  NEGVKAAW  TRA       01       10       01       29    DV5\n",
       "2     CAAKAGGGNKLTF  NEGVKAAW  TRA       01       10       01       29    DV5\n",
       "3  CAAITGIGGSQGNLIF  NEGVKAAW  TRA       01       42       01       13     01\n",
       "4    CAGTDGGATNKLIF  NEGVKAAW  TRA       01       32       01       36    DV7"
      ]
     },
     "execution_count": 465,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdjdb_data = vdjdb_data[sorted(vdjdb_data.columns)]\n",
    "vdjdb_data.head()\n",
    "# nog vragen: als er geen V_gene is gegeven, moet dit dan aangevuld worden tot 00 of gewoon leeggelaten worden?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enkel de epitopen overhouden die 30 keer of meer voorkomen,\n",
    "vdjdb_data.drop_duplicates\n",
    "vdjdb_pos = vdjdb_data.groupby('Epitope').filter(lambda x: len(x)> 30) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 467,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NEGVKAAW</th>\n",
       "      <th>KLGGALQAK</th>\n",
       "      <th>ELAGIGILTV</th>\n",
       "      <th>IVTDFSVIK</th>\n",
       "      <th>AVFDRKSDAK</th>\n",
       "      <th>RAKFKQLL</th>\n",
       "      <th>GILGFVFTL</th>\n",
       "      <th>SLFNTVATLY</th>\n",
       "      <th>RLRAEAQVK</th>\n",
       "      <th>AYAQKIFKI</th>\n",
       "      <th>GLCTLVAML</th>\n",
       "      <th>LLFGYPVYV</th>\n",
       "      <th>FLYALALLL</th>\n",
       "      <th>FLASKIGRLV</th>\n",
       "      <th>RTLNAWVKV</th>\n",
       "      <th>FLRGRAYGL</th>\n",
       "      <th>NLVPMVATV</th>\n",
       "      <th>CINGVCWTV</th>\n",
       "      <th>LLWNGPMAV</th>\n",
       "      <th>KLVALGINAV</th>\n",
       "      <th>DPFRLLQNSQVFS</th>\n",
       "      <th>GLIYNRMGAVTTEV</th>\n",
       "      <th>PKYVKQNTLKLAT</th>\n",
       "      <th>QARQMVQAMRTIGTHP</th>\n",
       "      <th>SGPLKAEIAQRLED</th>\n",
       "      <th>GMFNMLSTVLGVS</th>\n",
       "      <th>FRDYVDRFYKTLRAEQASQE</th>\n",
       "      <th>LLLGIGILV</th>\n",
       "      <th>KAFSPEVIPMF</th>\n",
       "      <th>DATYQRTRALVR</th>\n",
       "      <th>YLQPRTFLL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>total TCRAs</th>\n",
       "      <td>301.000000</td>\n",
       "      <td>13401.000000</td>\n",
       "      <td>361.000000</td>\n",
       "      <td>710.000000</td>\n",
       "      <td>1687.000000</td>\n",
       "      <td>1213.000000</td>\n",
       "      <td>4472.000000</td>\n",
       "      <td>37.000000</td>\n",
       "      <td>413.000000</td>\n",
       "      <td>39.000000</td>\n",
       "      <td>311.000000</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>41.000000</td>\n",
       "      <td>2212.000000</td>\n",
       "      <td>74.000000</td>\n",
       "      <td>299.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>43.000000</td>\n",
       "      <td>128.000000</td>\n",
       "      <td>141.000000</td>\n",
       "      <td>111.000000</td>\n",
       "      <td>63.000000</td>\n",
       "      <td>72.000000</td>\n",
       "      <td>248.000000</td>\n",
       "      <td>376.000000</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>112.000000</td>\n",
       "      <td>325.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique CDR3</th>\n",
       "      <td>118.000000</td>\n",
       "      <td>11426.000000</td>\n",
       "      <td>332.000000</td>\n",
       "      <td>519.000000</td>\n",
       "      <td>1534.000000</td>\n",
       "      <td>642.000000</td>\n",
       "      <td>3004.000000</td>\n",
       "      <td>37.000000</td>\n",
       "      <td>390.000000</td>\n",
       "      <td>39.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>2093.000000</td>\n",
       "      <td>71.000000</td>\n",
       "      <td>247.000000</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>42.000000</td>\n",
       "      <td>119.000000</td>\n",
       "      <td>139.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>62.000000</td>\n",
       "      <td>65.000000</td>\n",
       "      <td>141.000000</td>\n",
       "      <td>345.000000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>88.000000</td>\n",
       "      <td>262.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDR3 diversity</th>\n",
       "      <td>6.159650</td>\n",
       "      <td>13.212909</td>\n",
       "      <td>8.318918</td>\n",
       "      <td>8.044908</td>\n",
       "      <td>10.433803</td>\n",
       "      <td>8.097860</td>\n",
       "      <td>10.818358</td>\n",
       "      <td>5.209453</td>\n",
       "      <td>8.563207</td>\n",
       "      <td>5.285402</td>\n",
       "      <td>6.908826</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>3.812945</td>\n",
       "      <td>4.954196</td>\n",
       "      <td>5.459432</td>\n",
       "      <td>4.448181</td>\n",
       "      <td>10.948642</td>\n",
       "      <td>6.128372</td>\n",
       "      <td>7.743769</td>\n",
       "      <td>5.480084</td>\n",
       "      <td>5.379753</td>\n",
       "      <td>6.859375</td>\n",
       "      <td>7.111183</td>\n",
       "      <td>6.542164</td>\n",
       "      <td>5.945534</td>\n",
       "      <td>5.975481</td>\n",
       "      <td>6.660757</td>\n",
       "      <td>8.380361</td>\n",
       "      <td>4.324849</td>\n",
       "      <td>6.322849</td>\n",
       "      <td>7.817597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique V family</th>\n",
       "      <td>26.000000</td>\n",
       "      <td>33.000000</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>33.000000</td>\n",
       "      <td>33.000000</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>18.000000</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>33.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>23.000000</td>\n",
       "      <td>27.000000</td>\n",
       "      <td>18.000000</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>27.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>23.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique J family</th>\n",
       "      <td>38.000000</td>\n",
       "      <td>48.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>47.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>48.000000</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>43.000000</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>27.000000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>34.000000</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>37.000000</td>\n",
       "      <td>42.000000</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>34.000000</td>\n",
       "      <td>27.000000</td>\n",
       "      <td>43.000000</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>35.000000</td>\n",
       "      <td>37.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique V gene</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>8.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique J allele</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique V allele</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V family diversity</th>\n",
       "      <td>3.832061</td>\n",
       "      <td>4.471825</td>\n",
       "      <td>1.239394</td>\n",
       "      <td>4.247684</td>\n",
       "      <td>4.420575</td>\n",
       "      <td>3.745162</td>\n",
       "      <td>4.119155</td>\n",
       "      <td>3.677804</td>\n",
       "      <td>4.478998</td>\n",
       "      <td>3.759511</td>\n",
       "      <td>3.014938</td>\n",
       "      <td>3.237301</td>\n",
       "      <td>2.034864</td>\n",
       "      <td>3.619226</td>\n",
       "      <td>4.388536</td>\n",
       "      <td>3.389675</td>\n",
       "      <td>4.576203</td>\n",
       "      <td>2.495234</td>\n",
       "      <td>2.271051</td>\n",
       "      <td>1.864106</td>\n",
       "      <td>3.631444</td>\n",
       "      <td>3.540532</td>\n",
       "      <td>4.075117</td>\n",
       "      <td>2.856583</td>\n",
       "      <td>2.830259</td>\n",
       "      <td>3.308998</td>\n",
       "      <td>1.011571</td>\n",
       "      <td>2.641717</td>\n",
       "      <td>2.137972</td>\n",
       "      <td>3.478539</td>\n",
       "      <td>1.219611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>J family diversity</th>\n",
       "      <td>4.477435</td>\n",
       "      <td>5.390525</td>\n",
       "      <td>5.013560</td>\n",
       "      <td>5.007671</td>\n",
       "      <td>5.392408</td>\n",
       "      <td>4.711714</td>\n",
       "      <td>4.246334</td>\n",
       "      <td>4.465946</td>\n",
       "      <td>5.320349</td>\n",
       "      <td>4.355539</td>\n",
       "      <td>4.594717</td>\n",
       "      <td>4.515320</td>\n",
       "      <td>1.926574</td>\n",
       "      <td>4.518397</td>\n",
       "      <td>4.600922</td>\n",
       "      <td>4.051482</td>\n",
       "      <td>5.384800</td>\n",
       "      <td>4.762461</td>\n",
       "      <td>4.615085</td>\n",
       "      <td>4.246282</td>\n",
       "      <td>4.166246</td>\n",
       "      <td>4.827183</td>\n",
       "      <td>5.110287</td>\n",
       "      <td>4.383797</td>\n",
       "      <td>4.677688</td>\n",
       "      <td>4.657510</td>\n",
       "      <td>3.357188</td>\n",
       "      <td>4.834619</td>\n",
       "      <td>2.382494</td>\n",
       "      <td>4.711158</td>\n",
       "      <td>3.825526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V gene diversity</th>\n",
       "      <td>1.886072</td>\n",
       "      <td>2.537718</td>\n",
       "      <td>1.063756</td>\n",
       "      <td>2.273727</td>\n",
       "      <td>2.596383</td>\n",
       "      <td>2.412352</td>\n",
       "      <td>2.240680</td>\n",
       "      <td>2.427436</td>\n",
       "      <td>2.682693</td>\n",
       "      <td>2.338880</td>\n",
       "      <td>1.815277</td>\n",
       "      <td>2.516127</td>\n",
       "      <td>0.932138</td>\n",
       "      <td>2.411982</td>\n",
       "      <td>2.374000</td>\n",
       "      <td>2.150122</td>\n",
       "      <td>2.392475</td>\n",
       "      <td>1.843268</td>\n",
       "      <td>1.691508</td>\n",
       "      <td>1.576169</td>\n",
       "      <td>2.607912</td>\n",
       "      <td>2.083415</td>\n",
       "      <td>2.728008</td>\n",
       "      <td>2.201981</td>\n",
       "      <td>2.038810</td>\n",
       "      <td>2.165562</td>\n",
       "      <td>0.515611</td>\n",
       "      <td>2.296908</td>\n",
       "      <td>1.991864</td>\n",
       "      <td>1.913666</td>\n",
       "      <td>1.178457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>J allele diversity</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.008039</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.056070</td>\n",
       "      <td>0.200622</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.067670</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V allele diversity</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.024864</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      NEGVKAAW     KLGGALQAK  ELAGIGILTV   IVTDFSVIK  \\\n",
       "total TCRAs         301.000000  13401.000000  361.000000  710.000000   \n",
       "unique CDR3         118.000000  11426.000000  332.000000  519.000000   \n",
       "CDR3 diversity        6.159650     13.212909    8.318918    8.044908   \n",
       "unique V family      26.000000     33.000000   24.000000   31.000000   \n",
       "unique J family      38.000000     48.000000   46.000000   46.000000   \n",
       "unique V gene        10.000000     11.000000    9.000000   11.000000   \n",
       "unique J allele       1.000000      1.000000    1.000000    1.000000   \n",
       "unique V allele       1.000000      1.000000    1.000000    1.000000   \n",
       "V family diversity    3.832061      4.471825    1.239394    4.247684   \n",
       "J family diversity    4.477435      5.390525    5.013560    5.007671   \n",
       "V gene diversity      1.886072      2.537718    1.063756    2.273727   \n",
       "J allele diversity   -0.000000     -0.000000   -0.000000   -0.000000   \n",
       "V allele diversity   -0.000000     -0.000000   -0.000000   -0.000000   \n",
       "\n",
       "                     AVFDRKSDAK     RAKFKQLL    GILGFVFTL  SLFNTVATLY  \\\n",
       "total TCRAs         1687.000000  1213.000000  4472.000000   37.000000   \n",
       "unique CDR3         1534.000000   642.000000  3004.000000   37.000000   \n",
       "CDR3 diversity        10.433803     8.097860    10.818358    5.209453   \n",
       "unique V family       32.000000    33.000000    33.000000   16.000000   \n",
       "unique J family       47.000000    46.000000    48.000000   25.000000   \n",
       "unique V gene         11.000000    11.000000    11.000000    8.000000   \n",
       "unique J allele        1.000000     1.000000     2.000000    1.000000   \n",
       "unique V allele        1.000000     1.000000     2.000000    1.000000   \n",
       "V family diversity     4.420575     3.745162     4.119155    3.677804   \n",
       "J family diversity     5.392408     4.711714     4.246334    4.465946   \n",
       "V gene diversity       2.596383     2.412352     2.240680    2.427436   \n",
       "J allele diversity    -0.000000    -0.000000     0.008039   -0.000000   \n",
       "V allele diversity    -0.000000    -0.000000     0.024864   -0.000000   \n",
       "\n",
       "                     RLRAEAQVK  AYAQKIFKI   GLCTLVAML  LLFGYPVYV  FLYALALLL  \\\n",
       "total TCRAs         413.000000  39.000000  311.000000  32.000000  38.000000   \n",
       "unique CDR3         390.000000  39.000000  190.000000  32.000000  22.000000   \n",
       "CDR3 diversity        8.563207   5.285402    6.908826   5.000000   3.812945   \n",
       "unique V family      32.000000  18.000000   25.000000  13.000000  11.000000   \n",
       "unique J family      46.000000  24.000000   43.000000  25.000000  11.000000   \n",
       "unique V gene        11.000000   8.000000   10.000000   8.000000   5.000000   \n",
       "unique J allele       1.000000   1.000000    2.000000   2.000000   1.000000   \n",
       "unique V allele       1.000000   1.000000    1.000000   1.000000   1.000000   \n",
       "V family diversity    4.478998   3.759511    3.014938   3.237301   2.034864   \n",
       "J family diversity    5.320349   4.355539    4.594717   4.515320   1.926574   \n",
       "V gene diversity      2.682693   2.338880    1.815277   2.516127   0.932138   \n",
       "J allele diversity   -0.000000  -0.000000    0.056070   0.200622  -0.000000   \n",
       "V allele diversity   -0.000000  -0.000000   -0.000000  -0.000000  -0.000000   \n",
       "\n",
       "                    FLASKIGRLV  RTLNAWVKV  FLRGRAYGL    NLVPMVATV  CINGVCWTV  \\\n",
       "total TCRAs          31.000000  44.000000  41.000000  2212.000000  74.000000   \n",
       "unique CDR3          31.000000  44.000000  29.000000  2093.000000  71.000000   \n",
       "CDR3 diversity        4.954196   5.459432   4.448181    10.948642   6.128372   \n",
       "unique V family      14.000000  25.000000  15.000000    33.000000  11.000000   \n",
       "unique J family      25.000000  27.000000  22.000000    46.000000  34.000000   \n",
       "unique V gene         7.000000  10.000000   7.000000    11.000000   8.000000   \n",
       "unique J allele       1.000000   1.000000   1.000000     1.000000   1.000000   \n",
       "unique V allele       1.000000   1.000000   1.000000     1.000000   1.000000   \n",
       "V family diversity    3.619226   4.388536   3.389675     4.576203   2.495234   \n",
       "J family diversity    4.518397   4.600922   4.051482     5.384800   4.762461   \n",
       "V gene diversity      2.411982   2.374000   2.150122     2.392475   1.843268   \n",
       "J allele diversity   -0.000000  -0.000000  -0.000000    -0.000000  -0.000000   \n",
       "V allele diversity   -0.000000  -0.000000  -0.000000    -0.000000  -0.000000   \n",
       "\n",
       "                     LLWNGPMAV  KLVALGINAV  DPFRLLQNSQVFS  GLIYNRMGAVTTEV  \\\n",
       "total TCRAs         299.000000   46.000000      43.000000      128.000000   \n",
       "unique CDR3         247.000000   45.000000      42.000000      119.000000   \n",
       "CDR3 diversity        7.743769    5.480084       5.379753        6.859375   \n",
       "unique V family      28.000000   10.000000      17.000000       23.000000   \n",
       "unique J family      44.000000   25.000000      22.000000       37.000000   \n",
       "unique V gene        10.000000    7.000000       7.000000       10.000000   \n",
       "unique J allele       1.000000    1.000000       1.000000        1.000000   \n",
       "unique V allele       1.000000    1.000000       1.000000        1.000000   \n",
       "V family diversity    2.271051    1.864106       3.631444        3.540532   \n",
       "J family diversity    4.615085    4.246282       4.166246        4.827183   \n",
       "V gene diversity      1.691508    1.576169       2.607912        2.083415   \n",
       "J allele diversity   -0.000000   -0.000000      -0.000000       -0.000000   \n",
       "V allele diversity   -0.000000   -0.000000      -0.000000       -0.000000   \n",
       "\n",
       "                    PKYVKQNTLKLAT  QARQMVQAMRTIGTHP  SGPLKAEIAQRLED  \\\n",
       "total TCRAs            141.000000        111.000000       63.000000   \n",
       "unique CDR3            139.000000        102.000000       62.000000   \n",
       "CDR3 diversity           7.111183          6.542164        5.945534   \n",
       "unique V family         27.000000         18.000000       17.000000   \n",
       "unique J family         42.000000         31.000000       31.000000   \n",
       "unique V gene           10.000000          9.000000        6.000000   \n",
       "unique J allele          1.000000          1.000000        1.000000   \n",
       "unique V allele          1.000000          1.000000        1.000000   \n",
       "V family diversity       4.075117          2.856583        2.830259   \n",
       "J family diversity       5.110287          4.383797        4.677688   \n",
       "V gene diversity         2.728008          2.201981        2.038810   \n",
       "J allele diversity      -0.000000         -0.000000       -0.000000   \n",
       "V allele diversity      -0.000000         -0.000000       -0.000000   \n",
       "\n",
       "                    GMFNMLSTVLGVS  FRDYVDRFYKTLRAEQASQE   LLLGIGILV  \\\n",
       "total TCRAs             72.000000            248.000000  376.000000   \n",
       "unique CDR3             65.000000            141.000000  345.000000   \n",
       "CDR3 diversity           5.975481              6.660757    8.380361   \n",
       "unique V family         17.000000             10.000000   27.000000   \n",
       "unique J family         34.000000             27.000000   43.000000   \n",
       "unique V gene            8.000000              3.000000   11.000000   \n",
       "unique J allele          1.000000              2.000000    1.000000   \n",
       "unique V allele          1.000000              1.000000    1.000000   \n",
       "V family diversity       3.308998              1.011571    2.641717   \n",
       "J family diversity       4.657510              3.357188    4.834619   \n",
       "V gene diversity         2.165562              0.515611    2.296908   \n",
       "J allele diversity      -0.000000              0.067670   -0.000000   \n",
       "V allele diversity      -0.000000             -0.000000   -0.000000   \n",
       "\n",
       "                    KAFSPEVIPMF  DATYQRTRALVR   YLQPRTFLL  \n",
       "total TCRAs           31.000000    112.000000  325.000000  \n",
       "unique CDR3           22.000000     88.000000  262.000000  \n",
       "CDR3 diversity         4.324849      6.322849    7.817597  \n",
       "unique V family        7.000000     20.000000   23.000000  \n",
       "unique J family        9.000000     35.000000   37.000000  \n",
       "unique V gene          6.000000      7.000000    8.000000  \n",
       "unique J allele        1.000000      1.000000    1.000000  \n",
       "unique V allele        1.000000      1.000000    1.000000  \n",
       "V family diversity     2.137972      3.478539    1.219611  \n",
       "J family diversity     2.382494      4.711158    3.825526  \n",
       "V gene diversity       1.991864      1.913666    1.178457  \n",
       "J allele diversity    -0.000000     -0.000000   -0.000000  \n",
       "V allele diversity    -0.000000     -0.000000   -0.000000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Data exploration (gekopieerd uit de code van de paper + aangevuld voor de allelen)\n",
    "classes = vdjdb_pos['Epitope'].unique()\n",
    "stats = collections.defaultdict(list)\n",
    "for peptide in classes:\n",
    "    pep_data = vdjdb_pos[vdjdb_pos['Epitope'] == peptide]\n",
    "    \n",
    "    stats['total TCRAs'].append(pep_data.shape[0])\n",
    "    stats['unique CDR3'].append(len(set(pep_data['CDR3'])))\n",
    "    \n",
    "    cdr3_count = list(collections.Counter(pep_data['CDR3']).values())\n",
    "    stats['CDR3 diversity'].append(shannon(cdr3_count))\n",
    "    \n",
    "    stats['unique V family'].append(len(pep_data['V_family'].unique()))\n",
    "    stats['unique J family'].append(len(pep_data['J_family'].unique()))\n",
    "    stats['unique V gene'].append(len(pep_data['V_gene'].unique()))\n",
    "    stats['unique J allele'].append(len(pep_data['J_allele'].unique()))\n",
    "    stats['unique V allele'].append(len(pep_data['V_allele'].unique()))\n",
    "    v_fam_count = list(collections.Counter(pep_data['V_family']).values())\n",
    "    j_fam_count = list(collections.Counter(pep_data['J_family']).values())\n",
    "    v_gene_count = list(collections.Counter(pep_data['V_gene']).values())\n",
    "    j_allele_count = list(collections.Counter(pep_data['J_allele']).values())\n",
    "    v_allele_count = list(collections.Counter(pep_data['V_allele']).values())\n",
    "    stats['V family diversity'].append(shannon(v_fam_count))\n",
    "    stats['J family diversity'].append(shannon(j_fam_count))\n",
    "    stats['V gene diversity'].append(shannon(v_gene_count))\n",
    "    stats['J allele diversity'].append(shannon(j_allele_count))\n",
    "    stats['V allele diversity'].append(shannon(v_allele_count))\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "IPython.display.display(pd.DataFrame(stats, index=classes, columns=stats.keys()).T)\n",
    "# Unique J en V allele zeggen hier niet zo veel omdat deze bijna altijd alleen 01 of 02 zijn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Zelfde voor de data van McPAS\n",
    "McPAS_data = pd.read_csv(\"McPAS_data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3</th>\n",
       "      <th>Species</th>\n",
       "      <th>Epitope gene</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>MHC A</th>\n",
       "      <th>V</th>\n",
       "      <th>J</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAVTIGFGNVLHCGSGTQVIVLPHIQ</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ35</td>\n",
       "      <td>TRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASGGGADGLTFPYIQF</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>TRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAASPPESGGYNKLIF</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ4</td>\n",
       "      <td>TRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAAYYGGSOGNLIF</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>TRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAVSRGGGADGLTF</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>TRA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         CDR3 Species    Epitope gene     Epitope     MHC A  \\\n",
       "0  CAVTIGFGNVLHCGSGTQVIVLPHIQ   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "1           CASGGGADGLTFPYIQF   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "2            CAASPPESGGYNKLIF   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "3              CAAYYGGSOGNLIF   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "4              CAVSRGGGADGLTF   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "\n",
       "        V       J Gene  \n",
       "0   TRAV2  TRAJ35  TRA  \n",
       "1   TRAV2  TRAJ45  TRA  \n",
       "2  TRAV21   TRAJ4  TRA  \n",
       "3  TRAV21  TRAJ42  TRA  \n",
       "4   TRAV2  TRAJ45  TRA  "
      ]
     },
     "execution_count": 469,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del McPAS_data['Unnamed: 0']\n",
    "McPAS_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 470,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TRAJ35' 'TRAJ45' 'TRAJ4' 'TRAJ42' 'TRAJ47' 'TRAJ9' 'TRAJ39' 'TRAJ48'\n",
      " 'TRAJ1' 'TRAJ5-1' 'TRAJ15' nan 'TRAJ2' 'TRAJ5' 'TRAJ49' 'TRAJ54' 'TRAJ21'\n",
      " 'TRAJ53' 'TRAJ40' 'TRAJ36' 'TRAJ38' 'TRAJ26' 'TRAJ3' 'TRAJ37' 'TRAJ11'\n",
      " 'TRAJ31' 'TRAJ34' 'TRAJ23' 'TRAJ16' 'TRAJ1-3' 'TRAJ9-1' 'TRAJ16-5'\n",
      " 'TRAJ1-8' 'TRAJ10-1' 'TRAJ3-2' 'TRAJ9-14' 'TRAJ14-1' 'TRAJ16-1' 'TRAJ33'\n",
      " 'TRAJ52' 'TRAJ44' 'TRAJ43' 'TRAJ50' 'TRAJ20' 'TRAJ58' 'TRAJ6' 'TRAJ57'\n",
      " 'TRAJ8' 'TRAJ13' 'TRAJ29' 'TRAJ27' 'TRAJ30' 'TRAJ32' 'TRAJ10' 'TRAJ24'\n",
      " 'TRAJ7' 'TRAJ28' 'TRAJ56' 'TRAJ37-2' 'TRAJ41' 'TRAJ17' 'TRAJ22' 'TRAJ12'\n",
      " 'TRAJ4-01' 'TRAJ5-01' 'TRAJ3-01' 'TRAJ1-01' 'TRAJ24:02' 'TRAJ2-01'\n",
      " 'TRAJ9-01' 'TRAJ6-01' 'TRAJ8-01' 'TRAJ13:02' 'TRAJ49:01' 'TRAJ20:01'\n",
      " 'TRAJ42:01' 'TRAJ24:01' 'TRAJ21:01' 'TRAJ7:01' 'TRAJ39:01' 'TRAJ36:01'\n",
      " 'TRAJ54:01' 'TRAJ16:01' 'TRAJ58:01' 'TRAJ18' 'TRAJ53:01' 'TRAJ53:02'\n",
      " 'TRAJ53:05' 'TRAJ43:01' 'TRAJ2-1' 'TRAJ41:01' 'TRAJ3-1' 'TRAJ34:01'\n",
      " 'TRAJ31:01' 'TRAJ44:01' 'TRAJ38:01' 'TRAJ17:01' 'TRAJ48:01' 'TRAJ9:01'\n",
      " 'TRAJ18:01' 'TRAJ52:01' 'TRAJ56:01' 'TRAJ45:01' 'TRAJ3:01' 'TRAJ11:01'\n",
      " 'TRAJ27:01' 'TRAJ57:01' 'TRAJ6:01' 'TRAJ32:02' 'TRAJ22:01' 'TRAJ10:01'\n",
      " 'TRAJ32:01' 'TRAJ5:01' 'TRAJ37:02' 'TRAJ26:01' 'TRAJ47-1' 'TRAJ20-1'\n",
      " 'TRAJ24-1' 'TRAJ32-1' 'TRAJ44-1' 'TRAJ4-1' 'TRAJ37-1' 'TRAJ53-1'\n",
      " 'TRAJ34-1' 'TRAJ21-1' 'TRAJ33-1' 'TRAJ45-1' 'TRAJ54-1' 'TRAJ17-1'\n",
      " 'TRAJ42-1' 'TRAJ28-1' 'TRAJ40-1' 'TRAJ31-1']\n"
     ]
    }
   ],
   "source": [
    "TRAJ = McPAS_data ['J'].unique()\n",
    "TRAV = McPAS_data ['V'].unique()\n",
    "print (TRAJ) # gebruikt niet de standaard naamgeving\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TRAV2' 'TRAV21' 'TRAV10' 'TRAV8' 'TRAV8-2' 'TRAV3' 'TRAV9' 'TRAV1-1'\n",
      " 'TRAV25' 'TRAV1-4' 'TRAV1-2' 'TRAV28' 'TRAV16' 'TRAV2-1' 'TRAV26'\n",
      " 'TRAV39' 'TRAV15' 'TRAV20' 'TRAV12-1' 'TRAV2-3' 'TRAV4-1' 'TRAV15-1'\n",
      " 'TRAV22' 'TRAV26-2' 'TRAV8-3' 'TRAV18' 'TRAV4' 'TRAV23' 'TRAV27' 'TRAV19'\n",
      " 'TRAV2-2' 'TRAV6' 'TRAV12' nan 'TRAV5' 'TRAV14' 'TRAV12-4' 'TRAV38-2'\n",
      " 'TRAV24' 'TRAV17' 'TRAV38-1' 'TRAV29' 'TRAV9-2' 'TRAV31' 'TRAV1' 'TRAV11'\n",
      " 'TRAV1-01' 'TRAV13-1' 'TRAV8-4' 'TRAV8-6' 'TRAV251' 'TRAV40' 'TRAV12-2'\n",
      " 'TRAV21-1' 'TRAV13-2' 'TRAV8-1' 'TRAV26-1' 'TRAV21-2' 'TRAV3-1'\n",
      " 'TRAV35-1' 'TRAV29/DV5' 'TRAV14/DV4' 'TRAV23/DV6' 'TRAV12-2, TRAV21'\n",
      " 'TRAV12-3' 'TRAV2-01' 'TRAV35:01' 'TRAV12-3:01' 'TRAV25:01' 'TRAV3-01'\n",
      " 'TRAV10-01' 'TRAV1-2:01' 'TRAV14/DV4:01' 'TRAV29/DV5:01' 'TRAV1-1:01'\n",
      " 'TRAV6-01' 'TRAV38-2/DV8:01' 'TRAV13-2:01' 'TRAV5-01' 'TRAV23/DV6:01'\n",
      " 'TRAV9-2:01' 'TRAV41:01' 'TRAV38-1:01' 'TRAV14/DV4:02' 'TRAV8-1:01'\n",
      " 'TRAV12-2:01' 'TRAV22:01' 'TRAV8-6:02' 'TRAV8-6:01' 'TRAV8-3:01'\n",
      " 'TRAV13-1:01' 'TRAV41-01' 'TRAV8-2:01' 'TRAV21-01' 'TRAV26-2:01'\n",
      " 'TRAV8-4:01' 'TRAV4:01' 'TRAV16:01' 'TRAV36/DV7:01' 'TRAV12-1:01'\n",
      " 'TRAV26-1:01' 'TRAV16-01' 'TRAV34' 'TRAV35' 'TRAV38-2/DV8' 'TRAV41'\n",
      " 'TRAV24:01' 'TRAV5:01' 'TRAV3:01' 'TRAV20:02' 'TRAV6:02' 'TRAV19:01'\n",
      " 'TRAV30' 'TRAV36/DV7' 'TRAV7' 'TRAV29/DV5:02' 'TRAV29/DV5:05' 'TRAV17:01'\n",
      " 'TRAV6:03' 'TRAV26-1:03' 'TRAV27:01' 'TRAV24:01/TRAV24:02' 'TRAV24:01,'\n",
      " 'TRAV24:02' 'TRAV3 F' 'TRAV1-2\\xa0' 'TRAV22 F' 'TRAV09-2' 'TRAV25 F'\n",
      " 'TRAV38-2/DV8 F' 'TRAV121' 'TRAV-2' 'TRAV08-3' 'TRAV08-2' 'TRAV08-4'\n",
      " 'TRAV01-2' 'TRAV08-1' 'TRAV38-2/DV8*01' 'TRAV26-1*01' 'TRAV24*01'\n",
      " 'TRAV39*01' 'TRAV4*01' 'TRAV12-2*01' 'TRAV17*01' 'TRAV38-1*01'\n",
      " 'TRAV3*01 F' 'TRAV41*01' 'TRAV8-6*01' 'TRAV9-2*01' 'TRAV1-2*01'\n",
      " 'TRAV14/DV4*01' 'TRAV13-2*01' 'TRAV16*01' 'TRA1-2' 'mTRDV2-2' 'TRAV9-1'\n",
      " 'mTRAV14D-1']\n"
     ]
    }
   ],
   "source": [
    "print (TRAV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "metadata": {},
   "outputs": [],
   "source": [
    "McPAS_data.replace('TRAJ','',regex=True, inplace = True)\n",
    "McPAS_data[['J_family','J_allele_1']] = McPAS_data.J.str.split (\"-\",expand=True)\n",
    "McPAS_data[['J_family','J_allele_2']] = McPAS_data.J_family.str.split (\":\",expand=True)\n",
    "#McPAS_data.rename(index={J: 'J_family'})\n",
    "del McPAS_data ['J']\n",
    "McPAS_data.replace(np.nan,'', regex=True )\n",
    "# ten slotte de 2 gene kolommen nog mergen en de andere 2 verwijderen\n",
    "McPAS_data = McPAS_data.assign(J_allele = McPAS_data.J_allele_1.astype(str) + McPAS_data.J_allele_2.astype(str))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "metadata": {},
   "outputs": [],
   "source": [
    "McPAS_data.replace('nan','',regex=True, inplace = True)\n",
    "del McPAS_data ['J_allele_1']\n",
    "del McPAS_data ['J_allele_2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3</th>\n",
       "      <th>Species</th>\n",
       "      <th>Epitope gene</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>MHC A</th>\n",
       "      <th>V</th>\n",
       "      <th>Gene</th>\n",
       "      <th>J_family</th>\n",
       "      <th>J_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAVTIGFGNVLHCGSGTQVIVLPHIQ</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRA</td>\n",
       "      <td>35</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASGGGADGLTFPYIQF</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRA</td>\n",
       "      <td>45</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAASPPESGGYNKLIF</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRA</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAAYYGGSOGNLIF</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRA</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAVSRGGGADGLTF</td>\n",
       "      <td>Human</td>\n",
       "      <td>Melan-A/MART-1</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRA</td>\n",
       "      <td>45</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         CDR3 Species    Epitope gene     Epitope     MHC A  \\\n",
       "0  CAVTIGFGNVLHCGSGTQVIVLPHIQ   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "1           CASGGGADGLTFPYIQF   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "2            CAASPPESGGYNKLIF   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "3              CAAYYGGSOGNLIF   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "4              CAVSRGGGADGLTF   Human  Melan-A/MART-1  EAAGIGILTV  HLA-A*02   \n",
       "\n",
       "        V Gene J_family J_allele  \n",
       "0   TRAV2  TRA       35           \n",
       "1   TRAV2  TRA       45           \n",
       "2  TRAV21  TRA        4           \n",
       "3  TRAV21  TRA       42           \n",
       "4   TRAV2  TRA       45           "
      ]
     },
     "execution_count": 474,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "McPAS_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "metadata": {},
   "outputs": [],
   "source": [
    "#McPAS_data.replace('TRAV','',regex=True, inplace = True)\n",
    "#McPAS_data[['V','V_Fct']] = McPAS_data.V.str.split (\" \",expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 476,
   "metadata": {},
   "outputs": [],
   "source": [
    "McPAS_pos = McPAS_data.groupby('Epitope').filter(lambda x: len(x)> 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "metadata": {},
   "outputs": [],
   "source": [
    "# negative data\n",
    "# nog andere data zoeken want deze heeft overal als allele 00 staan, \n",
    "\n",
    "# ImmuneACCESSdb : filters set control, human, geen TCRA data beschikbaar\n",
    "# iReceptor: filters set case-control, human, TRA, en daarna gekeken welke controle data ongeveer de correcte grootte had.\n",
    "neg_data = pd.read_csv(\"neg_data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 478,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>v_call</th>\n",
       "      <th>j_call</th>\n",
       "      <th>locus</th>\n",
       "      <th>junction_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV38-2DV8*00</td>\n",
       "      <td>TRAJ47*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAYWEYGNKLVF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV27*00</td>\n",
       "      <td>TRAJ22*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAGAATSGSARQLTF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV20*00</td>\n",
       "      <td>TRAJ6*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVQAGSYIPTF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV13-2*00</td>\n",
       "      <td>TRAJ16*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVLTG_GQKLLF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV1-1*00</td>\n",
       "      <td>TRAJ26*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVTPQGQNFVF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequence          v_call     j_call locus      junction_aa\n",
       "0       NaN  TRAV38-2DV8*00  TRAJ47*00   TRA     CAYWEYGNKLVF\n",
       "1       NaN       TRAV27*00  TRAJ22*00   TRA  CAGAATSGSARQLTF\n",
       "2       NaN       TRAV20*00   TRAJ6*00   TRA     CAVQAGSYIPTF\n",
       "3       NaN     TRAV13-2*00  TRAJ16*00   TRA    CAVLTG_GQKLLF\n",
       "4       NaN      TRAV1-1*00  TRAJ26*00   TRA     CAVTPQGQNFVF"
      ]
     },
     "execution_count": 478,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del neg_data ['Unnamed: 0']\n",
    "neg_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 479,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>v_call</th>\n",
       "      <th>locus</th>\n",
       "      <th>junction_aa</th>\n",
       "      <th>J_family</th>\n",
       "      <th>J_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV38-2DV8*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAYWEYGNKLVF</td>\n",
       "      <td>47</td>\n",
       "      <td>00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV27*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAGAATSGSARQLTF</td>\n",
       "      <td>22</td>\n",
       "      <td>00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV20*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVQAGSYIPTF</td>\n",
       "      <td>06</td>\n",
       "      <td>00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV13-2*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVLTG_GQKLLF</td>\n",
       "      <td>16</td>\n",
       "      <td>00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV1-1*00</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVTPQGQNFVF</td>\n",
       "      <td>26</td>\n",
       "      <td>00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequence          v_call locus      junction_aa J_family J_allele\n",
       "0       NaN  TRAV38-2DV8*00   TRA     CAYWEYGNKLVF       47       00\n",
       "1       NaN       TRAV27*00   TRA  CAGAATSGSARQLTF       22       00\n",
       "2       NaN       TRAV20*00   TRA     CAVQAGSYIPTF       06       00\n",
       "3       NaN     TRAV13-2*00   TRA    CAVLTG_GQKLLF       16       00\n",
       "4       NaN      TRAV1-1*00   TRA     CAVTPQGQNFVF       26       00"
      ]
     },
     "execution_count": 479,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neg_data.replace('TRAJ','',regex=True, inplace = True)\n",
    "neg_data[['J_family','J_allele']] = neg_data.j_call.str.split (\"*\",expand=True)\n",
    "del neg_data ['j_call']\n",
    "# add a leading 0 to all families with only one digit\n",
    "neg_data['J_family'] = neg_data['J_family'].str.zfill(2)\n",
    "neg_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_data.replace('TRAV','',regex=True, inplace = True)\n",
    "neg_data.replace('TRDV','',regex=True, inplace = True)\n",
    "neg_data[['v_call','V_allele']] = neg_data.v_call.str.split (\"*\",expand=True) # eerst allele afsplitsen, komt na de *\n",
    "# dan in 3 stappen de gene van de family scheiden\n",
    "neg_data[['V_family','V_gene']] = neg_data.v_call.str.split (\"-\",expand=True) # eerst scheiden op -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 481,
   "metadata": {},
   "outputs": [],
   "source": [
    "del neg_data ['v_call']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 482,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a leading 0 to all Values with only one digit\n",
    "neg_data['V_family'] = neg_data['V_family'].str.zfill(2)\n",
    "neg_data['V_gene'] = neg_data['V_gene'].str.zfill(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 483,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>J_allele</th>\n",
       "      <th>J_family</th>\n",
       "      <th>V_allele</th>\n",
       "      <th>V_family</th>\n",
       "      <th>V_gene</th>\n",
       "      <th>junction_aa</th>\n",
       "      <th>locus</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00</td>\n",
       "      <td>47</td>\n",
       "      <td>00</td>\n",
       "      <td>38</td>\n",
       "      <td>2DV8</td>\n",
       "      <td>CAYWEYGNKLVF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00</td>\n",
       "      <td>22</td>\n",
       "      <td>00</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>CAGAATSGSARQLTF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00</td>\n",
       "      <td>06</td>\n",
       "      <td>00</td>\n",
       "      <td>20</td>\n",
       "      <td>None</td>\n",
       "      <td>CAVQAGSYIPTF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>00</td>\n",
       "      <td>16</td>\n",
       "      <td>00</td>\n",
       "      <td>13</td>\n",
       "      <td>02</td>\n",
       "      <td>CAVLTG_GQKLLF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>00</td>\n",
       "      <td>26</td>\n",
       "      <td>00</td>\n",
       "      <td>01</td>\n",
       "      <td>01</td>\n",
       "      <td>CAVTPQGQNFVF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  J_allele J_family V_allele V_family V_gene      junction_aa locus  sequence\n",
       "0       00       47       00       38   2DV8     CAYWEYGNKLVF   TRA       NaN\n",
       "1       00       22       00       27   None  CAGAATSGSARQLTF   TRA       NaN\n",
       "2       00       06       00       20   None     CAVQAGSYIPTF   TRA       NaN\n",
       "3       00       16       00       13     02    CAVLTG_GQKLLF   TRA       NaN\n",
       "4       00       26       00       01     01     CAVTPQGQNFVF   TRA       NaN"
      ]
     },
     "execution_count": 483,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neg_data = neg_data[sorted(neg_data.columns)]\n",
    "neg_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Feature generation\n",
    "# gekopieerd uit de code van de paper\n",
    "\n",
    "# physico-chemical amino acid properties\n",
    "# = dictionaries\n",
    "basicity = {'A': 206.4, 'B': 210.7, 'C': 206.2, 'D': 208.6, 'E': 215.6, 'F': 212.1, 'G': 202.7,\n",
    "            'H': 223.7, 'I': 210.8, 'K': 221.8, 'L': 209.6, 'M': 213.3, 'N': 212.8, 'P': 214.4,\n",
    "            'Q': 214.2, 'R': 237.0, 'S': 207.6, 'T': 211.7, 'V': 208.7, 'W': 216.1, 'X': 210.2,\n",
    "            'Y': 213.1, 'Z': 214.9}\n",
    "\n",
    "hydrophobicity = {'A': 0.16, 'B': -3.14, 'C': 2.50, 'D': -2.49, 'E': -1.50, 'F': 5.00, 'G': -3.31,\n",
    "                  'H': -4.63, 'I': 4.41, 'K': -5.00, 'L': 4.76, 'M': 3.23, 'N': -3.79, 'P': -4.92,\n",
    "                  'Q': -2.76, 'R': -2.77, 'S': -2.85, 'T': -1.08, 'V': 3.02, 'W': 4.88, 'X': 4.59,\n",
    "                  'Y': 2.00, 'Z': -2.13}\n",
    "\n",
    "helicity = {'A': 1.24, 'B': 0.92, 'C': 0.79, 'D': 0.89, 'E': 0.85, 'F': 1.26, 'G': 1.15, 'H': 0.97,\n",
    "            'I': 1.29, 'K': 0.88, 'L': 1.28, 'M': 1.22, 'N': 0.94, 'P': 0.57, 'Q': 0.96, 'R': 0.95,\n",
    "            'S': 1.00, 'T': 1.09, 'V': 1.27, 'W': 1.07, 'X': 1.29, 'Y': 1.11, 'Z': 0.91}\n",
    "\n",
    "mutation_stability = {'A': 13, 'C': 52, 'D': 11, 'E': 12, 'F': 32, 'G': 27, 'H': 15, 'I': 10,\n",
    "                      'K': 24, 'L': 34, 'M':  6, 'N':  6, 'P': 20, 'Q': 10, 'R': 17, 'S': 10,\n",
    "                      'T': 11, 'V': 17, 'W': 55, 'Y': 31}\n",
    "\n",
    "physchem_properties = {'basicity': basicity, 'hydrophobicity': hydrophobicity,\n",
    "                       'helicity': helicity, 'mutation stability': mutation_stability}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:RL]",
   "language": "python",
   "name": "conda-env-RL-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
